### PR TITLE
Complete object graph and unique'ify all objects

### LINKED ISSUE #107: Modellable Child <-> Parent Relationship Creates Endless Loop
> - contentful.swift version number: 0.8.0
> - Xcode version number: 9.0 +
> - Target operating system(s) and version number(s)
    - [X] iOS: 
    - [X] tvOS: 
    - [X] watchOS:
    - [X] macOS:
> - Package manager:
    - [X] Carthage
    - [X] Cocoapods

It appears when I have a parent that contains an array of children, and the children reference the parent, during the mapping process there is an infinite loop that causes the framework to crash. The child will try to map the parent, then the parent re-attempts at mapping the child which then causes the child try to map the parent again, and so on..... until the crash.

In my code I am currently mapping my data manually until this is fixed, but hopefully it inspires a solution on your side to fix it.  For this example imagine a Series, which contains Seasons, which contains Videos. I may need each entity at some point to reference it's parent. Thus the Video would need to tell me which season it is part of, and each season may need to tell me which series it is part of. 

```
final class Series {
    static let contentTypeId: String = "series"
    var seasons : [Season]?

    init?(entry: Entry) {
        ...
        if let seasonLinks = entry.fields["seasons"] as? [Link] {
            var seasonArray = [Season]()
            
            for link in seasonLinks {
                switch link {
                case .entry(let seasonEntry):
                    if let season = Season(entry: seasonEntry, parentSeries : self) {
                        seasonArray.append(season)
                    }
                default:
                    continue
                }
            }
            
            seasons = seasonArray
        }   
    }
}

final class Season {
    static let contentTypeId: String = "season"
    var parentSeason : Series?
   
    init?(entry: Entry, parentSeries : Series? = nil) {
       ...
        if parentSeries != nil {
            self.parentSeries = parentSeries
        } else {
            // parse the parent
        }
    }
}

final class Video  {
    static let contentTypeId: String = "video"
    var parentSeason : Season?
   
    init?(entry: Entry, parentSeason : Season? = nil) {
        ...
        if parentSeries != nil {
            self.parentSeries = parentSeries
        } else {
            // parse the parent
        }
    }
}

```

If there is a way to identify the parent relationship, and you can tell the child "Hey, I'm your elder don't map me", the child can mapping it and stop the infinite loop. hopefully this inspires some creativity on your part to identify these relationships during mapping :)

### PR DESCRIPTION
**Pull request by loudmouth**
Fixes #107

### CODE REVIEW

File: Sources/Client+Modellable.swift [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    public let entries: [ContentTypeId: [EntryModellable]]
+}
+
+
+extension Client {
+    /**
+     Fetch a collection of Entries from Contentful matching the specified query.
+
+     - Parameter query: The Query object to match results againts.
+     - Returns: An Observable forr the resulting `MappedContent` container.
```
Comment (mariobodemann) [inline Sources/Client+Modellable.swift]: `forr` vs `for`

File: Sources/Client+Modellable.swift [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    @discardableResult public func fetchMappedEntries(with query: Query) -> Observable<Result<MappedContent>> {
+        let asyncDataTask: AsyncDataTask<Query, MappedContent> = fetchMappedEntries(with:then:)
+        return toObservable(parameter: query, asyncDataTask: asyncDataTask).observable
+    }
+
+    /**
+     Fetches all entries and includes matching the passed in `Query`. The completion handler returned will return a `MappedContent` object which
+     contains an array of `Asset`s and a dictionary of ContentTypeId's to arrays of `EntryModellable` types of your own defining.
```
Comment (mariobodemann) [inline Sources/Client+Modellable.swift]: `defining` vs `definition`.

File: Sources/ContentModellable.swift [hunk 3]
```diff
@@ -10,41 +10,212 @@ import Foundation
...
 
-    init(spaceType: SpaceModellable.Type, entryTypes: [EntryModellable.Type]) {
-        self.spaceType = spaceType
+    /// An array of model class types defined in your application which will be returned when using the relevant fetch methods on `Client`.
+    public let entryTypes: [EntryModellable.Type]
+
+    /**
+     Initializes a new `ContentModel` instance.
+     
+     - Parameter entryTypes: References to the the types of your own defition, conforming to `EntryModellable`, which
```
Comment (mariobodemann) [inline Sources/ContentModellable.swift]: `defition` vs `definition`

File: Sources/ContentModellable.swift [hunk 4]
```diff
@@ -10,41 +10,212 @@ import Foundation
...
+        cachedRelationshipNames[entryType.contentTypeId] = relationshipNames
+        return relationshipNames
+    }
+
+    // MARK: Relationships.
+
+    // A type used to cache relationships that should be deleted in the `resolveRelationships()` method.
+    fileprivate struct DeletedRelationship {}
+
+    // Returns a dictionary representing the fields names and the target id(s) to create links to.
```
Comment (mariobodemann) [inline Sources/ContentModellable.swift]: `fields names` or should it be `field names`?

File: Sources/DataCache.swift [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    public static func cacheKey(for resource: EntryModellable) -> String {
+        let delimeter = DataCache.cacheKeyDelimiter
+
+        let cacheKey =  resource.id + delimeter + "entry" + delimeter + resource.localeCode
+        return cacheKey
+    }
+
+    public static func cacheKey(for resource: LocalizableResource) -> String {
+        let delimeter = "_"
```
Comment (mariobodemann) [inline Sources/DataCache.swift]: hmm, I am seeing double here. Do you mind using the `let constant` of the `DataCache` here?

Comment (loudmouth) [inline Sources/DataCache.swift]: good cache! :-P

**Pull request merged**